src/shogun/preprocessor/RandomFourierGaussPreproc.h

```diff
@@ -170,68 +142,49 @@ class RandomFourierGaussPreproc: public RandomMixin<DensePreprocessor<float64_t>
 	/// return a type of preprocessor
 	virtual EPreprocessorType get_type() const { return P_RANDOMFOURIERGAUSS; }
 
+private:
+	void init();
+
 protected:
 	/** default processing routine, inherited from base class
 	 * @param matrix the features matrix to be processed
 	 * @return the processed feature matrix from the DenseFeatures<float64_t>
 	 * class in case (2) (see description above) this routine requires only
 	 * steps 2a) and 2b), the rest is determined automatically
 	 */
-	virtual SGMatrix<float64_t> apply_to_matrix(SGMatrix<float64_t> matrix);
+	virtual SGMatrix<float64_t> apply_to_matrix(const SGMatrix<float64_t>& matrix);
```

Comment:
I'm a little sceptical about this being a protected virtual method rather than a public virtual method. Thoughts?

Reply:
If we agree on that matter, we'd need another PR to fix DensePreprocessor and its derivatives, which I'd be happy to make ;)

Reply:
where are these methods called from?

Reply:
It looks like a helper function, for example in DensePreprocessor we have the function ```transform``` which does some checks before passing it to ```apply_to_matrix``` method. 

```c++
transform(){
	require(features->get_feature_class()==C_DENSE, "Provided features ({}) "
			"has to be of C_DENSE ({}) class!",
			features->get_feature_class(), C_DENSE);
        ....
        auto feat_matrix = apply_to_matrix(matrix);
        return std::make_shared<DenseFeatures<ST>>(feat_matrix);
}
```
It's defined virtual, so we can override it here. I guess we'd need to define transform class and put apply_to_matrix inside this function, right?

Reply:
I think all is good then, no? Keep in mind the base interface is what will be eventually be exposed to the user, not the particular class interface.

Reply:
Agreed, what about the arg type though? It's pass by value rather than pass by const reference.

Reply:
definitely change, although it doesnt make a huge difference in this case (as this method is only called once and call-by-value for SGMatrix/Vector doesnt copy the memory, but only the wrapping class structure (check the code)

Reply:
Now that I think of it, it's better if its call-by-value. Now, one has the luxury to pass both lvalue and rvalue to the args and since the size isn't that big(since it's a wrapper to the pointer) it won't be expensive. We'll keep it!. Thanks for clarifying.

Reply:
in general, we want to avoid call by value if it is not necessary, but in particular for things that are called in loops